src/transformers/models/dab_detr/modeling_dab_detr.py

```diff
@@ -0,0 +1,2329 @@
+            if self.training:
+                dropout_probability = torch.rand([])
+                if dropout_probability < self.layerdrop:
+                    continue
+
+            obj_center = reference_points[..., : self.config.query_dim]
+            query_sine_embed = gen_sine_position_embeddings(obj_center, self.d_model)
+            query_pos = self.ref_point_head(query_sine_embed)
+
+            # For the first decoder layer, we do not apply transformation over p_s
+            if self.query_scale_type != "fix_elewise":
+                if layer_id == 0:
+                    pos_transformation = 1
+                else:
+                    pos_transformation = self.query_scale(hidden_states)
+            else:
+                pos_transformation = self.query_scale.weight[layer_id]
+
+            # apply transformation
+            query_sine_embed = query_sine_embed[..., : self.config.d_model] * pos_transformation
+
+            # modulated HW attentions
+            if self.config.decoder_modulate_hw_attn:
+                refHW_cond = self.ref_anchor_head(hidden_states).sigmoid()  # nq, bs, 2
+                query_sine_embed[..., self.d_model // 2 :] *= (refHW_cond[..., 0] / obj_center[..., 2]).unsqueeze(-1)
+                query_sine_embed[..., : self.d_model // 2] *= (refHW_cond[..., 1] / obj_center[..., 3]).unsqueeze(-1)
+
+            layer_outputs = decoder_layer(
+                hidden_states,
+                attention_mask=None,
+                object_queries=object_queries,
+                query_position_embeddings=query_pos,
+                query_sine_embed=query_sine_embed,
+                encoder_hidden_states=encoder_hidden_states,
+                encoder_attention_mask=memory_key_padding_mask,
+                output_attentions=output_attentions,
+            )
+
+            # iter update
+            hidden_states = layer_outputs[0]
+
+            if self.bbox_embed is not None:
+                if self.decoder_bbox_embed_diff_each_layer:
+                    tmp = self.bbox_embed[layer_id](hidden_states)
+                else:
+                    tmp = self.bbox_embed(hidden_states)
+
+                tmp[..., : self.config.query_dim] += inverse_sigmoid(reference_points)
+                new_reference_points = tmp[..., : self.config.query_dim].sigmoid()
+                if layer_id != self.num_layers - 1:
+                    ref_points.append(new_reference_points)
+                reference_points = new_reference_points.detach()
+
+            intermediate.append(self.layernorm(hidden_states))
+
+            if output_attentions:
+                all_self_attns += (layer_outputs[1],)
+
+                if encoder_hidden_states is not None:
+                    all_cross_attentions += (layer_outputs[2],)
+
+        if self.layernorm is not None:
+            hidden_states = self.layernorm(hidden_states)
+            intermediate.pop()
+            intermediate.append(hidden_states)
```

Comment:
I tried to simplify as much as possible:
- Layerdrop has been deleted.
- Code paths have been removed.
- _query_scale_type_ has been simplified since we only use _cond_elewise_.
- _decoder_modulate_hw_attn_ has been deleted because all pre-trained versions set this variable to true.
- Gradient checkpointing has been added.
- The temporary variable has been renamed to new_reference_point, why it was labeled as tmp, no clue.

Reply:
great effort, thanks 🤗